### DISCUSSION TITLE
Welcome to CNApy Discussions!

### DISCUSSION THREAD

**Discussion by sthiele** [cnapy-org/CNApy#90] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by galib96** [comment]
Hi, I am a grad student and using cnapy GUI for Elementary flux modes analysis. 

Is it possible to use cnapy inside jupyter notebook so that generated flux data can be imported/exported and analyzed in python?

**Reply by Paulocracy** [reply]
Hello Galib!
It's certainly possible to start CNApy inside a Jupyter notebook by using cnapy.py's content in a Jupyter cell, however, as of now, there wouldn't be much use to it as CNApy doesn't generate much output on its own.
However, with the latest CNApy version 1.1.6, you can save the reaction list (including shown fluxes for a specific EFM analysis) concent to your system's clipboard by right-clicking on the list's column captions. Alternatively, you can also directly store this list as a CSV file or Excel XLSX under the "Clipboard" menu entry. These files (or clipboard content) can then be loaded into Python using e.g. pandas, csv or openpyxl.
If you have further suggestions or questions feel free to post them, we're always looking for feedback and ideas :-)

**Reply by axelvonkamp** [reply]
For elementary model analysis in particular you can also use the efmtool-link package independent of CNApy. It has an example Jupyter notebook efmtool_link.ipynb included.

**Comment by kelseyaw** [comment]
Subject: Clarification on Units for Exchange Reaction Constraints in CNAPy

I am working on a flux balance analysis model using CNAPy and have a question regarding unit consistency. The biological uptake rates I have are based on experimental data and are in units of 𝜇mol/gDW/min resulting in values that are quite small (e.g., −0.0009 to 0.0037
etc.). To my understanding, these uptake rates should be applied as constraints by adjusting the bounds of the corresponding exchange reactions. However, the model itself appears to represent reaction bounds with default values of −1000 to 1000 for unlimited flow. If this is derived from cobrapy, which, as I understand, operates in units of mmol/gDW/hr, I would like to confirm:

1) Is cobrapy’s default unit for reaction flux bounds indeed mmol/gDW/hr?
2) What is the recommended approach to apply my experimental constraints? Should I convert the 𝜇mol/gDW/min data to mmol/gDW/hr (e.g., by multiplying by 60/1000) before adjusting the bounds of the exchange reactions? Because this would result in even smaller numbers which will not be biological relevant given the unlimited bounds of -1000 to 1000. 

I would appreciate any advice / guidance!

**Reply by axelvonkamp** [reply]
Hi, the unit for reaction fluxes is not really defined anywhere but using mmol/(gDW * hr) is the most common approach. You could also use any other unit, you just need to be careful how the growth function is defined. The metabolites consumed by the growth function usually sum up to 1 gDW biomass and the coefficients usually have the unit mmol/gDW. With such a growth function you need to use mmol/(gDW * [time unit]) for the fluxes to be consistent, the growth rate is then in 1/[time unit]. Of course, you can adapt the growth function so that it works for alternative flux units. Since you have very small fluxes it may be advisable to switch to a 𝜇mol/gDW growth function and use 𝜇mol/(gDW * hr) for fluxes to avoid potential numerical issues but I would try with mmol/(gDW * hr) first and see if it works.

**Comment by PeaceCris** [comment]
Hi,

I’m really interested in the thermodynamic approaches you’ve developed, such as optMDFpathway. I’d like to know if there’s a way to use the algorithm directly on an interactive computing platform, like Jupyter, for example.

I’ve been using make scenario feasible directly in Jupyter, and I must say it’s amazing, I truly appreciate your work.

Thank you for everything, and I hope you have a great day!

**Reply by axelvonkamp** [reply]
Hi, check out this package: https://github.com/klamt-lab/COBRA-k
